lib/elixir/lib/calendar/date.ex

```diff
@@ -123,6 +134,20 @@ defmodule Date do
     raise ArgumentError, "both dates must have matching calendars"
   end
 
+  def range(%{calendar: calendar} = first, %Duration{} = duration) do
+    last = shift(first, duration)
+    range(first, last)
+  end
+
+  def range(%{calendar: calendar} = first, duration) when is_list(duration) do
+    last = shift(first, duration)
+    range(first, last)
+  end
+
+  def range(%{calendar: _, year: _, month: _, day: _}, _duration) do
+    raise ArgumentError, "expected a date or duration as second argument"
+  end
```

Comment:
@josevalim do we usually do this, or just let this fall through and raise a `FunctionClauseError`? I think this kind of thing would be caught by types anyway